go: enums not working - Expected enum value, got "<string>"

## :bug: Bug Report

### Affected Languages
<!--
Check the box (with an X) for any language runtime that you know is affected by
the reported bug. If you're uncertain whether a language is affected or not,
please leave the bux un-checked.
-->
- [x] `Golang`
- [ ] `TypeScript` or `Javascript`
- [ ] `Python`
- [ ] `Java`
- [ ] .NET (`C#`, `F#`, ...)

### General Information
* **JSII Version:** 1.20.0
* **Platform:**     *any*

### What is the problem?

Enums are not working.

Consider the following jsii code packaged as `github.com/eladb/go-enum-repro/enumrepro`:

```ts
export enum MyEnum {
  MY_VALUE = 'my',
  YOUR_VALUE = 'your'
}

export class Hello {
  public sayHello(x: MyEnum) {
    return `hello, ${x}!`;
  }
}
```

And the consuming go code:

```go
package main

import (
	"github.com/eladb/go-enum-repro/enumrepro"
)

func main() {
	var hello = enumrepro.NewHello()
	hello.SayHello(enumrepro.MyEnumYourValue)
}
```

When running this program, we get the following error:

```
panic: "Expected enum value, got \"YOUR_VALUE\""

goroutine 1 [running]:
github.com/aws/jsii-runtime-go.Invoke(0x1124220, 0x1376d60, 0x11461c8, 0x8, 0xc00015bdf0, 0x1, 0x1, 0x1376d01, 0x111c400, 0xc000104720, ...)
        /Users/benisrae/go/pkg/mod/github.com/aws/jsii-runtime-go@v1.20.0/runtime.go:87 +0x305
github.com/eladb/go-enum-repro/enumrepro.(*Hello).SayHello(0x1376d60, 0x11466ca, 0xa, 0x111e4a0, 0xc000102058)
        /Users/benisrae/playground/play-0204133257/enum-repro/dist/go/enumrepro/enumrepro.go:34 +0x165
main.main()
        /Users/benisrae/playground/play-0204133257/enum-repro/dist/go/test/main.go:9 +0x4d
```
- [ ] blocked by https://github.com/aws/jsii/issues/2561

--- a/gh-pages/content/specification/2-type-system.md
+++ b/gh-pages/content/specification/2-type-system.md
@@ -230,10 +230,10 @@ value:
 
 In **JavaScript**, `enum` entries are represented by their value equivalent. In order to support statically typed
 representations in other languages, these are serialized using a dedicated wrapper object, using a single key
-(`$jsii.enum`) with the fully qualified name of the `enum` entry:
+(`$jsii.enum`) with the fully qualified name of the `enum` entry (formatted as `<enum type fqn>/<entry name>`):
 
 ```json
-{ "$jsii.enum": "@scope/module.EnumType.ENTRY_NAME" }
+{ "$jsii.enum": "@scope/module.EnumType/ENTRY_NAME" }
 ```
 
 ### Identity Serialization

--- a/packages/@jsii/go-runtime/jsii-calc-test/main_test.go
+++ b/packages/@jsii/go-runtime/jsii-calc-test/main_test.go
@@ -157,6 +157,46 @@ func TestAllTypes(t *testing.T) {
 	})
 }
 
+func TestEnumUnmarshal(t *testing.T) {
+	actual := calc.EnumDispenser_RandomStringLikeEnum()
+	if actual != calc.StringEnumB {
+		t.Errorf("Expected StringEnum.B. Actual: %s", actual)
+	}
+}
+
+func TestEnumRoundtrip(t *testing.T) {
+	allTypes := calc.NewAllTypes()
+	actual := allTypes.EnumMethod(calc.StringEnumA)
+	if actual != calc.StringEnumA {
+		t.Errorf("Expected StringEnum.A. Actual: %s", actual)
+	}
+
+	actual = allTypes.EnumMethod(calc.StringEnumC)
+	if actual != calc.StringEnumC {
+		t.Errorf("Expected StringEnum.C. Actual: %s", actual)
+	}
+}
+
+func TestOptionalEnums(t *testing.T) {
+	allTypes := calc.NewAllTypes()
+	actual := allTypes.GetOptionalEnumValue()
+	if actual != "" {
+		t.Error("Expected value to be nil")
+	}
+
+	allTypes.SetOptionalEnumValue(calc.StringEnumB)
+	actual = allTypes.GetOptionalEnumValue()
+	if actual != calc.StringEnumB {
+		t.Errorf("Expected StringEnum.B. Actual: %s", actual)
+	}
+
+	allTypes.SetOptionalEnumValue("")
+	actual = allTypes.GetOptionalEnumValue()
+	if actual != "" {
+		t.Error("Expected value to be nil")
+	}
+}
+
 func TestReturnsSpecialParam(t *testing.T) {
 	retSpecialParam := returnsParam.NewReturnsSpecialParameter()
 	val := retSpecialParam.ReturnsSpecialParam()

--- a/packages/@jsii/go-runtime/jsii-runtime-go/api.go
+++ b/packages/@jsii/go-runtime/jsii-runtime-go/api.go
@@ -89,6 +89,10 @@ type objref struct {
 	InstanceID string `json:"$jsii.byref"`
 }
 
+type enumref struct {
+	MemberFQN string `json:"$jsii.enum"`
+}
+
 type loadRequest struct {
 	kernelRequester
 

--- a/packages/@jsii/go-runtime/jsii-runtime-go/api.go
+++ b/packages/@jsii/go-runtime/jsii-runtime-go/api.go
@@ -85,11 +85,11 @@ func (r kernelResponder) isResponse() {
 	return
 }
 
-type objref struct {
+type objectRef struct {
 	InstanceID string `json:"$jsii.byref"`
 }
 
-type enumref struct {
+type enumRef struct {
 	MemberFQN string `json:"$jsii.enum"`
 }
 
@@ -130,7 +130,7 @@ type delRequest struct {
 	kernelRequester
 
 	API    string `json:"api"`
-	ObjRef objref `json:"objref"`
+	ObjRef objectRef `json:"objref"`
 }
 
 type delResponse struct {
@@ -142,7 +142,7 @@ type getRequest struct {
 
 	API      string `json:"api"`
 	Property string `json:"property"`
-	ObjRef   objref `json:"objref"`
+	ObjRef   objectRef `json:"objref"`
 }
 
 type staticGetRequest struct {
@@ -165,7 +165,7 @@ type setRequest struct {
 	API      string      `json:"api"`
 	Property string      `json:"property"`
 	Value    interface{} `json:"value"`
-	ObjRef   objref      `json:"objref"`
+	ObjRef   objectRef      `json:"objref"`
 }
 
 type staticSetRequest struct {
@@ -196,7 +196,7 @@ type invokeRequest struct {
 	API       string        `json:"api"`
 	Method    string        `json:"method"`
 	Arguments []interface{} `json:"args"`
-	ObjRef    objref        `json:"objref"`
+	ObjRef    objectRef        `json:"objref"`
 }
 
 type invokeResponse struct {
@@ -211,7 +211,7 @@ type beginRequest struct {
 	API       string        `json:"api"`
 	Method    *string       `json:"method"`
 	Arguments []interface{} `json:"args"`
-	ObjRef    objref        `json:"objref"`
+	ObjRef    objectRef        `json:"objref"`
 }
 
 type beginResponse struct {

--- a/packages/@jsii/go-runtime/jsii-runtime-go/runtime.go
+++ b/packages/@jsii/go-runtime/jsii-runtime-go/runtime.go
@@ -117,7 +117,7 @@ func Invoke(obj interface{}, method string, args []interface{}, hasReturn bool, 
 		API:       "invoke",
 		Method:    method,
 		Arguments: castPtrsToRef(args),
-		ObjRef: objref{
+		ObjRef: objectRef{
 			InstanceID: refid,
 		},
 	})
@@ -167,7 +167,7 @@ func Get(obj interface{}, property string, ret interface{}) {
 	res, err := client.get(getRequest{
 		API:      "get",
 		Property: property,
-		ObjRef: objref{
+		ObjRef: objectRef{
 			InstanceID: refid,
 		},
 	})
@@ -213,7 +213,7 @@ func Set(obj interface{}, property string, value interface{}) {
 		API:      "set",
 		Property: property,
 		Value:    castPtrToRef(value),
-		ObjRef: objref{
+		ObjRef: objectRef{
 			InstanceID: refid,
 		},
 	})
@@ -240,8 +240,8 @@ func StaticSet(fqn FQN, property string, value interface{}) {
 	}
 }
 
-func castValToRef(data interface{}) (objref, bool) {
-	ref := objref{}
+func castValToRef(data interface{}) (objectRef, bool) {
+	ref := objectRef{}
 	ok := false
 	dataVal := reflect.ValueOf(data)
 
@@ -262,7 +262,7 @@ func castValToRef(data interface{}) (objref, bool) {
 	return ref, ok
 }
 
-func castValToEnumRef(data interface{}) (enum enumref, ok bool) {
+func castValToEnumRef(data interface{}) (enum enumRef, ok bool) {
 	dataVal := reflect.ValueOf(data)
 	ok = false
 
@@ -294,11 +294,10 @@ func castPtrToRef(data interface{}) interface{} {
 	if dataVal.Kind() == reflect.Ptr {
 		valref, valHasRef := client.findObjectRef(data)
 		if valHasRef {
-			return objref{InstanceID: valref}
+			return objectRef{InstanceID: valref}
 		}
 	} else if dataVal.Kind() == reflect.String {
-		enumRef, isEnumRef := client.types.tryRenderEnumRef(dataVal)
-		if isEnumRef {
+		if enumRef, isEnumRef := client.types.tryRenderEnumRef(dataVal); isEnumRef {
 			return enumRef
 		}
 	} else if dataVal.Kind() == reflect.Slice {
@@ -336,17 +335,16 @@ func (c *client) castAndSetToPtr(ptr interface{}, data interface{}) {
 	ref, isRef := castValToRef(data)
 	if isRef {
 		// If return data is JSII object references, add to objects table.
-		concreteType, err := c.types.concreteTypeFor(ptrVal.Type())
-		if err != nil {
+		if concreteType, err := c.types.concreteTypeFor(ptrVal.Type()); err == nil {
+			ptrVal.Set(reflect.New(concreteType))
+			c.objects[ptrVal.Interface()] = ref.InstanceID
+		} else {
 			panic(err)
 		}
-		ptrVal.Set(reflect.New(concreteType))
-		c.objects[ptrVal.Interface()] = ref.InstanceID
 		return
 	}
 
-	enumref, isEnum := castValToEnumRef(data)
-	if isEnum {
+	if enumref, isEnum := castValToEnumRef(data); isEnum {
 		member, err := c.types.enumMemberForEnumRef(enumref)
 		if err != nil {
 			panic(err)

--- a/packages/@jsii/go-runtime/jsii-runtime-go/runtime.go
+++ b/packages/@jsii/go-runtime/jsii-runtime-go/runtime.go
@@ -262,6 +262,27 @@ func castValToRef(data interface{}) (objref, bool) {
 	return ref, ok
 }
 
+func castValToEnumRef(data interface{}) (enum enumref, ok bool) {
+	dataVal := reflect.ValueOf(data)
+	ok = false
+
+	if dataVal.Kind() == reflect.Map {
+		for _, k := range dataVal.MapKeys() {
+			// Finding values type requires extracting from reflect.Value
+			// otherwise .Kind() returns `interface{}`
+			v := reflect.ValueOf(dataVal.MapIndex(k).Interface())
+
+			if k.Kind() == reflect.String && k.String() == "$jsii.enum" && v.Kind() == reflect.String {
+				enum.MemberFQN = v.String()
+				ok = true
+				return
+			}
+		}
+	}
+
+	return
+}
+
 // Accepts pointers to structs that implement interfaces and searches for an
 // existing object reference in the client. If it exists, it casts it to an
 // objref for the runtime. Recursively casts types that may contain nested
@@ -274,6 +295,11 @@ func castPtrToRef(data interface{}) interface{} {
 		valref, valHasRef := client.findObjectRef(data)
 		if valHasRef {
 			return objref{InstanceID: valref}
+		}
+	} else if dataVal.Kind() == reflect.String {
+		enumRef, isEnumRef := client.types.tryRenderEnumRef(dataVal)
+		if isEnumRef {
+			return enumRef
 		}
 	} else if dataVal.Kind() == reflect.Slice {
 		refs := make([]interface{}, dataVal.Len())
@@ -308,7 +334,29 @@ func (c *client) castAndSetToPtr(ptr interface{}, data interface{}) {
 	dataVal := reflect.ValueOf(data)
 
 	ref, isRef := castValToRef(data)
+	if isRef {
+		// If return data is JSII object references, add to objects table.
+		concreteType, err := c.types.concreteTypeFor(ptrVal.Type())
+		if err != nil {
+			panic(err)
+		}
+		ptrVal.Set(reflect.New(concreteType))
+		c.objects[ptrVal.Interface()] = ref.InstanceID
+		return
+	}
 
+	enumref, isEnum := castValToEnumRef(data)
+	if isEnum {
+		member, err := c.types.enumMemberForEnumRef(enumref)
+		if err != nil {
+			panic(err)
+		}
+
+		ptrVal.Set(reflect.ValueOf(member))
+		return
+	}
+
+	// arrays
 	if ptrVal.Kind() == reflect.Slice && dataVal.Kind() == reflect.Slice {
 		// If return type is a slice, recursively cast elements
 		for i := 0; i < dataVal.Len(); i++ {
@@ -318,15 +366,13 @@ func (c *client) castAndSetToPtr(ptr interface{}, data interface{}) {
 			c.castAndSetToPtr(inner.Interface(), dataVal.Index(i).Interface())
 			ptrVal.Set(reflect.Append(ptrVal, inner.Elem()))
 		}
-	} else if isRef {
-		// If return data is JSII object references, add to objects table.
-		concreteType, err := c.types.concreteTypeFor(ptrVal.Type())
-		if err != nil {
-			panic(err)
-		}
-		ptrVal.Set(reflect.New(concreteType))
-		c.objects[ptrVal.Interface()] = ref.InstanceID
-	} else {
+
+		return
+	}
+
+	// TODO: maps
+
+	if data != nil {
 		val := reflect.ValueOf(data)
 		ptrVal.Set(val)
 	}

--- a/packages/@jsii/go-runtime/jsii-runtime-go/type-registry.go
+++ b/packages/@jsii/go-runtime/jsii-runtime-go/type-registry.go
@@ -27,7 +27,7 @@ type typeRegistry struct {
 
 	// maps Go enum type ("StringEnum") to the corresponding jsii enum FQN (e.g.
 	// "jsii-calc.StringEnum")
-	typeToEnumFqn map[reflect.Type]FQN
+	typeToEnumFQN map[reflect.Type]FQN
 }
 
 // newTypeRegistry creates a new type registry.
@@ -36,7 +36,7 @@ func newTypeRegistry() *typeRegistry {
 		fqnToType:       make(map[FQN]reflect.Type),
 		ifaceToStruct:   make(map[reflect.Type]reflect.Type),
 		fqnToEnumMember: make(map[string]interface{}),
-		typeToEnumFqn:   make(map[reflect.Type]FQN),
+		typeToEnumFQN:   make(map[reflect.Type]FQN),
 	}
 }
 
@@ -72,22 +72,24 @@ func (t *typeRegistry) registerEnum(fqn FQN, enm reflect.Type, members map[strin
 	if enm.Kind() != reflect.String {
 		return fmt.Errorf("the provided enum is not a string derivative: %v", enm)
 	}
-
+	if existing, exists := t.fqnToType[fqn]; exists && existing != enm {
+		return fmt.Errorf("another type was already registered with %s: %v", fqn, existing)
+	}
 	for memberName, memberVal := range members {
 		vt := reflect.ValueOf(memberVal).Type()
 		if vt != enm {
 			return fmt.Errorf("the enum entry for key %s has incorrect type %v", memberName, vt)
 		}
-
-		memberFqn := fmt.Sprintf("%s/%s", fqn, memberName)
-		t.fqnToEnumMember[memberFqn] = memberVal
+		// Not setting in t.fqnToEnumMember here so we don't cause any side-effects
+		// if the pre-condition fails at any point. This is done in a second loop.
 	}
 
-	if existing, exists := t.fqnToType[fqn]; exists && existing != enm {
-		return fmt.Errorf("another type was already registered with %s: %v", fqn, existing)
+	t.typeToEnumFQN[enm] = fqn
+	for memberName, memberVal := range members {
+		memberFQN := fmt.Sprintf("%s/%s", fqn, memberName)
+		t.fqnToEnumMember[memberFQN] = memberVal
 	}
 
-	t.typeToEnumFqn[enm] = fqn
 	return nil
 }
 
@@ -150,8 +152,7 @@ func (t *typeRegistry) concreteTypeFor(typ reflect.Type) (structType reflect.Typ
 		err = nil
 	} else if typ.Kind() == reflect.Interface {
 		var ok bool
-		structType, ok = t.ifaceToStruct[typ]
-		if !ok {
+		if structType, ok = t.ifaceToStruct[typ]; !ok {
 			err = fmt.Errorf("no concrete type was registered for interface: %v", typ)
 		}
 	} else {
@@ -160,33 +161,36 @@ func (t *typeRegistry) concreteTypeFor(typ reflect.Type) (structType reflect.Typ
 	return
 }
 
-func (t *typeRegistry) enumMemberForEnumRef(ref enumref) (interface{}, error) {
-	member, ok := t.fqnToEnumMember[ref.MemberFQN]
-	if !ok {
-		return nil, fmt.Errorf("no enum member registered for %s", ref.MemberFQN)
+// enumMemberForEnumRef returns the go enum member corresponding to a jsii fully
+// qualified enum member name (e.g: "jsii-calc.StringEnum/A"). If no enum member
+// was registered (via registerEnum) for the provided enumref, an error is
+// returned.
+func (t *typeRegistry) enumMemberForEnumRef(ref enumRef) (interface{}, error) {
+	if member, ok := t.fqnToEnumMember[ref.MemberFQN]; ok {
+		return member, nil
 	}
-
-	return member, nil
+	return nil, fmt.Errorf("no enum member registered for %s", ref.MemberFQN)
 }
 
-func (t *typeRegistry) tryRenderEnumRef(value reflect.Value) (ref *enumref, isEnumRef bool) {
+// tryRenderEnumRef returns an enumref if the provided value corresponds to a
+// registered enum type. The returned enumref is nil if the provided enum value
+// is a zero-value (i.e: "").
+func (t *typeRegistry) tryRenderEnumRef(value reflect.Value) (ref *enumRef, isEnumRef bool) {
 	if value.Kind() != reflect.String {
 		isEnumRef = false
 		return
 	}
 
-	enumFqn, ok := t.typeToEnumFqn[value.Type()]
-	if !ok {
-		isEnumRef = false
-		return
-	}
-
-	isEnumRef = true
-
-	if memberName := value.String(); memberName != "" {
-		ref = &enumref{MemberFQN: fmt.Sprintf("%s/%s", enumFqn, memberName)}
+	if enumFQN, ok := t.typeToEnumFQN[value.Type()]; ok {
+		isEnumRef = true
+		if memberName := value.String(); memberName != "" {
+			ref = &enumRef{MemberFQN: fmt.Sprintf("%s/%s", enumFQN, memberName)}
+		} else {
+			ref = nil
+		}
 	} else {
-		ref = nil
+		isEnumRef = false
 	}
+
 	return
 }

--- a/packages/@jsii/go-runtime/jsii-runtime-go/type-registry.go
+++ b/packages/@jsii/go-runtime/jsii-runtime-go/type-registry.go
@@ -13,6 +13,7 @@ type typeRegistry struct {
 	// qualified type name. The kind of type being returned depends on what the
 	// FQN represents... This will be the second argument of provided to a
 	// register* function.
+	// enums are not included
 	fqnToType map[FQN]reflect.Type
 
 	// ifaceToStruct provides the go struct that should be used to build a proxy
@@ -20,17 +21,22 @@ type typeRegistry struct {
 	// conversion.
 	ifaceToStruct map[reflect.Type]reflect.Type
 
-	// enumMembers has enum constants for each registered enum type, supporting
-	// correct conversion of enum values received from JavaScript.
-	enumMembers map[reflect.Type]map[string]interface{}
+	// map enum member FQNs (e.g. "jsii-calc.StringEnum/A") to the corresponding
+	// go const for this member.
+	fqnToEnumMember map[string]interface{}
+
+	// maps Go enum type ("StringEnum") to the corresponding jsii enum FQN (e.g.
+	// "jsii-calc.StringEnum")
+	typeToEnumFqn map[reflect.Type]FQN
 }
 
 // newTypeRegistry creates a new type registry.
 func newTypeRegistry() *typeRegistry {
 	return &typeRegistry{
-		fqnToType:     make(map[FQN]reflect.Type),
-		ifaceToStruct: make(map[reflect.Type]reflect.Type),
-		enumMembers:   make(map[reflect.Type]map[string]interface{}),
+		fqnToType:       make(map[FQN]reflect.Type),
+		ifaceToStruct:   make(map[reflect.Type]reflect.Type),
+		fqnToEnumMember: make(map[string]interface{}),
+		typeToEnumFqn:   make(map[reflect.Type]FQN),
 	}
 }
 
@@ -67,20 +73,21 @@ func (t *typeRegistry) registerEnum(fqn FQN, enm reflect.Type, members map[strin
 		return fmt.Errorf("the provided enum is not a string derivative: %v", enm)
 	}
 
-	for k, v := range members {
-		vt := reflect.ValueOf(v).Type()
+	for memberName, memberVal := range members {
+		vt := reflect.ValueOf(memberVal).Type()
 		if vt != enm {
-			return fmt.Errorf("the enum entry for key %s has incorrect type %v", k, vt)
+			return fmt.Errorf("the enum entry for key %s has incorrect type %v", memberName, vt)
 		}
+
+		memberFqn := fmt.Sprintf("%s/%s", fqn, memberName)
+		t.fqnToEnumMember[memberFqn] = memberVal
 	}
 
 	if existing, exists := t.fqnToType[fqn]; exists && existing != enm {
 		return fmt.Errorf("another type was already registered with %s: %v", fqn, existing)
 	}
 
-	t.fqnToType[fqn] = enm
-	t.enumMembers[enm] = members
-
+	t.typeToEnumFqn[enm] = fqn
 	return nil
 }
 
@@ -149,6 +156,37 @@ func (t *typeRegistry) concreteTypeFor(typ reflect.Type) (structType reflect.Typ
 		}
 	} else {
 		err = fmt.Errorf("invalid argument: expected a struct or interface type, but got %v", typ)
+	}
+	return
+}
+
+func (t *typeRegistry) enumMemberForEnumRef(ref enumref) (interface{}, error) {
+	member, ok := t.fqnToEnumMember[ref.MemberFQN]
+	if !ok {
+		return nil, fmt.Errorf("no enum member registered for %s", ref.MemberFQN)
+	}
+
+	return member, nil
+}
+
+func (t *typeRegistry) tryRenderEnumRef(value reflect.Value) (ref *enumref, isEnumRef bool) {
+	if value.Kind() != reflect.String {
+		isEnumRef = false
+		return
+	}
+
+	enumFqn, ok := t.typeToEnumFqn[value.Type()]
+	if !ok {
+		isEnumRef = false
+		return
+	}
+
+	isEnumRef = true
+
+	if memberName := value.String(); memberName != "" {
+		ref = &enumref{MemberFQN: fmt.Sprintf("%s/%s", enumFqn, memberName)}
+	} else {
+		ref = nil
 	}
 	return
 }

